### PR TITLE
Create donor landing and handle of invest's origins

### PR DESCRIPTION
**Pull request by davidbeig**
- create donate landing action
- create invest origin model and repositories
- add styles for donate landing
- add fields to handle invest origin donations

### CODE REVIEW

File: db/migrations/20220429110112_goteo_invest_origin.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+   * @return string The SQL string to execute for the Up migration.
+   */
+  public function getUpSQL()
+  {
+      /**
+       *     invest_origin
+      source
+      detail
+      allocated
+       */
```
Comment (carlos-mg89) [inline db/migrations/20220429110112_goteo_invest_origin.php]: We could clear this up

File: src/Goteo/Entity/Invest/InvestOrigin.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+namespace Goteo\Entity\Invest;
+
+use Goteo\Model\Invest;
+
+class InvestOrigin
+{
+    private int $invest_id;
+    private string $source;
+    private string $detail;
+    private string $allocated;
```
Comment (carlos-mg89) [inline src/Goteo/Entity/Invest/InvestOrigin.php]: `$allocated` could be null, as the SQL doesn't state NOT NULL. So this property should be `?string`. Unless the SQL should be NOT NULL, in that case, the model would be already fine.

File: src/Goteo/Entity/Invest/InvestOrigin.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        return $this->detail;
+    }
+
+    public function setDetail(string $detail): InvestOrigin
+    {
+        $this->detail = $detail;
+        return $this;
+    }
+
+    public function getAllocated(): string
```
Comment (carlos-mg89) [inline src/Goteo/Entity/Invest/InvestOrigin.php]: If the property could be null, then this method should return ?string

File: src/Goteo/Entity/Invest/InvestOrigin.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        return $this->allocated;
+    }
+
+    public function setAllocated(string $allocated): InvestOrigin
```
Comment (carlos-mg89) [inline src/Goteo/Entity/Invest/InvestOrigin.php]: If the property could be null, then this method's parameter should be of ?string type

File: src/Goteo/Repository/InvestOriginRepository.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * @return string[]
+     */
+    public function getSources(): array {
+        $sql = "SELECT invest_origin.source
+                FROM invest_origin
+                GROUP BY invest_origin.source";
+
+        return $this->query($sql)->fetchAll();
+    }
```
Comment (carlos-mg89) [inline src/Goteo/Repository/InvestOriginRepository.php]: If source could have repeated values, maybe it'd be interesting to have a `SELECT DISTINCT(invest_origin.source)` instead.

File: src/Goteo/Repository/InvestOriginRepository.php [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * @return string[]
+     */
+    public function getDetails(): array {
+        $sql = "SELECT invest_origin.detail
+                FROM invest_origin
+                GROUP BY invest_origin.detail";
+
+        return $this->query($sql)->fetchAll();
+    }
```
Comment (carlos-mg89) [inline src/Goteo/Repository/InvestOriginRepository.php]: If detail could have repeated values, maybe it'd be interesting to have a `SELECT DISTINCT(invest_origin.detail)` instead.

File: src/Goteo/Repository/InvestOriginRepository.php [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * @return string[]
+     */
+    public function getAllocateds(): array {
+        $sql = "SELECT invest_origin.allocated
+                FROM invest_origin
+                GROUP BY invest_origin.allocated";
+
+        return $this->query($sql)->fetchAll();
+    }
```
Comment (carlos-mg89) [inline src/Goteo/Repository/InvestOriginRepository.php]: If allocated could have repeated values, maybe it'd be interesting to have a `SELECT DISTINCT(invest_origin.allocated)` instead.

File: tests/Goteo/Repository/InvestOriginRepositoryTest.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        $this->repository = new InvestOriginRepository();
+    }
+
+    public function testExceptionOnGetByInvestId(): void {
+        $id = 1;
+        $this->expectException(ModelNotFoundException::class);
+        $this->expectExceptionMessage("InvestOrigin with invest_id {$id} not found");
+
+        $this->repository->getByInvestId($id);
+
```
Comment (carlos-mg89) [inline tests/Goteo/Repository/InvestOriginRepositoryTest.php:27]: Could this line be removed please?

**Review by carlos-mg89 — CHANGES_REQUESTED** [review]
Needs more thorough testing (tests fail). Although I was able to see the donate page and seems to work as expected.

Some minor cleaning could be done (comments left).

Overall, looks quite nice. I'm sure it can be finished quickly.

File: tests/Goteo/Repository/InvestOriginRepositoryTest.php (hunk 8 above)
Comment (davidbeig) [reply tests/Goteo/Repository/InvestOriginRepositoryTest.php:27]: I think not... This test expects the ModelNotFound exception and the message written, so it has to execute the code that actually throws the esception.

As far as i understand, this is the correct way of doing this test.

Comment (carlos-mg89) [reply tests/Goteo/Repository/InvestOriginRepositoryTest.php:27]: I meant the empty line, line 27.

Comment (davidbeig) [reply tests/Goteo/Repository/InvestOriginRepositoryTest.php:27]: Ah ok! I though you meant the call to getByInvestId!

**Review by carlos-mg89 — DISMISSED** [review]

**Comment by carlos-mg89** [comment]
Now it looks great!

**Review by carlos-mg89 — APPROVED** [review]

**Pull request merged**